File: core-patch/src/main/scala/eu/ostrzyciel/jelly/core/patch/internal/PatchEncoderImpl.scala
package eu.ostrzyciel.jelly.core.patch.internal

import eu.ostrzyciel.jelly.core.JellyExceptions.RdfProtoSerializationError
import eu.ostrzyciel.jelly.core.internal.NodeEncoderFactory
import eu.ostrzyciel.jelly.core.patch.*
import eu.ostrzyciel.jelly.core.proto.v1.*
import eu.ostrzyciel.jelly.core.proto.v1.patch.*
import eu.ostrzyciel.jelly.core.{NodeEncoder, ProtoEncoderConverter}

import scala.annotation.experimental

/**
 * Implementation of PatchEncoder.
 * @param converter the converter to use
 * @param params parameters for the encoder
 * @tparam TNode the type of RDF nodes in the library
 */
@experimental
final class PatchEncoderImpl[TNode >: Null](
  protected val converter: ProtoEncoderConverter[TNode, ?, ?],
  params: PatchEncoder.Params,
) extends PatchEncoder[TNode]:

  override val options: RdfPatchOptions = params.options

  /**
   * In contrast to the ProtoEncoder, this buffer is ALWAYS externally managed.
   */
  private val rowBuffer = params.rowBuffer

  override protected val nodeEncoder: NodeEncoder[TNode] = NodeEncoderFactory.create[TNode](
    options.maxPrefixTableSize,
    options.maxNameTableSize,
    options.maxDatatypeTableSize,
    this,
  )
  private var emittedOptions: Boolean = false

  private[core] override def appendNameEntry(entry: RdfNameEntry): Unit =
    rowBuffer.append(RdfPatchRow.ofName(entry))

  private[core] override def appendPrefixEntry(entry: RdfPrefixEntry): Unit =
    rowBuffer.append(RdfPatchRow.ofPrefix(entry))

  private[core] override def appendDatatypeEntry(entry: RdfDatatypeEntry): Unit =
    rowBuffer.append(RdfPatchRow.ofDatatype(entry))

  override def addTriple(s: TNode, p: TNode, o: TNode): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofStatementAdd(tripleInQuadToProto(s, p, o)))

  override def deleteTriple(s: TNode, p: TNode, o: TNode): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofStatementDelete(tripleInQuadToProto(s, p, o)))

  override def addQuad(s: TNode, p: TNode, o: TNode, g: TNode): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofStatementAdd(quadToProto(s, p, o, g)))

  override def deleteQuad(s: TNode, p: TNode, o: TNode, g: TNode): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofStatementDelete(quadToProto(s, p, o, g)))

  override def transactionStart(): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofTransactionStart)

  override def transactionCommit(): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofTransactionCommit)

  override def transactionAbort(): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofTransactionAbort)

  override def addNamespace(name: String, iriValue: TNode, graph: TNode): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofNamespaceAdd(
      RdfPatchNamespace(
        name,
        converter.nodeToProto(nodeEncoder, iriValue).iri,
        encodeNsIri(graph),
      )
    ))

  override def deleteNamespace(name: String, iriValue: TNode, graph: TNode): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofNamespaceDelete(
      RdfPatchNamespace(
        name,
        encodeNsIri(iriValue),
        encodeNsIri(graph),
      )
    ))

  override def header(key: String, value: TNode): Unit =
    handleStreamStart()
    rowBuffer.append(RdfPatchRow.ofHeader(
      RdfPatchHeader(key, converter.nodeToProto(nodeEncoder, value))
    ))

  override def punctuation(): Unit =
    handleStreamStart()
    if !options.streamType.isPunctuated then
      throw new RdfProtoSerializationError("Punctuation is not allowed in this stream type.")
    rowBuffer.append(RdfPatchRow.ofPunctuation)
  
  private def encodeNsIri(iri: TNode): RdfIri =
    if iri == null then null
    else converter.nodeToProto(nodeEncoder, iri).iri
  
  private inline def handleStreamStart(): Unit =
    if !emittedOptions then emitOptions()

  private def emitOptions(): Unit =
    emittedOptions = true
    rowBuffer.append(RdfPatchRow.ofOptions(
      // Override whatever the user set in the options.
      options.withVersion(PatchConstants.protoVersion)
    ))

File: core-patch/src/main/java/eu/neverblink/jelly/core/patch/internal/PatchEncoderImpl.java
package eu.neverblink.jelly.core.patch.internal;

import eu.neverblink.jelly.core.*;
import eu.neverblink.jelly.core.patch.PatchEncoder;
import eu.neverblink.jelly.core.proto.v1.RdfDatatypeEntry;
import eu.neverblink.jelly.core.proto.v1.RdfNameEntry;
import eu.neverblink.jelly.core.proto.v1.RdfPrefixEntry;
import eu.neverblink.jelly.core.proto.v1.patch.*;

/**
 * Implementation of PatchEncoder.
 *
 * @param <TNode> the type of RDF nodes in the library
 */
@ExperimentalApi
@InternalApi
public class PatchEncoderImpl<TNode> extends PatchEncoder<TNode> {

    private boolean hasEmittedOptions = false;

    // These rows are always identical, so we can use singletons
    private static final RdfPatchRow ROW_PUNCTUATION = RdfPatchRow.newInstance()
        .setPunctuation(RdfPatchPunctuation.EMPTY);
    private static final RdfPatchRow ROW_TX_START = RdfPatchRow.newInstance()
        .setTransactionStart(RdfPatchTransactionStart.EMPTY);
    private static final RdfPatchRow ROW_TX_COMMIT = RdfPatchRow.newInstance()
        .setTransactionCommit(RdfPatchTransactionCommit.EMPTY);
    private static final RdfPatchRow ROW_TX_ABORT = RdfPatchRow.newInstance()
        .setTransactionAbort(RdfPatchTransactionAbort.EMPTY);

    static {
        // Pre-calculate the serialized sizes of the static rows to avoid repeated calculations
        ROW_PUNCTUATION.getSerializedSize();
        ROW_TX_START.getSerializedSize();
        ROW_TX_COMMIT.getSerializedSize();
        ROW_TX_ABORT.getSerializedSize();
    }

    /**
     * Constructor.
     *
     * @param converter the converter to use
     * @param params    parameters for the encoder
     */
    public PatchEncoderImpl(ProtoEncoderConverter<TNode> converter, PatchEncoder.Params params) {
        super(converter, params);
    }

    @Override
    public void appendNameEntry(RdfNameEntry nameEntry) {
        // Calculate the size of the row now, as all objects are likely still in L1/L2 cache.
        rowBuffer.appendMessage().setName(nameEntry).getSerializedSize();
    }

    @Override
    public void appendPrefixEntry(RdfPrefixEntry prefixEntry) {
        // Calculate the size of the row now, as all objects are likely still in L1/L2 cache.
        rowBuffer.appendMessage().setPrefix(prefixEntry).getSerializedSize();
    }

    @Override
    public void appendDatatypeEntry(RdfDatatypeEntry datatypeEntry) {
        rowBuffer.appendMessage().setDatatype(datatypeEntry).getSerializedSize();
    }

    @Override
    public void addQuad(TNode subject, TNode predicate, TNode object, TNode graph) {
        emitOptions();
        final var quad = quadToProto(subject, predicate, object, graph);
        rowBuffer.appendMessage().setStatementAdd(quad).getSerializedSize();
    }

    @Override
    public void deleteQuad(TNode subject, TNode predicate, TNode object, TNode graph) {
        emitOptions();
        final var quad = quadToProto(subject, predicate, object, graph);
        rowBuffer.appendMessage().setStatementDelete(quad).getSerializedSize();
    }

    @Override
    public void addTriple(TNode subject, TNode predicate, TNode object) {
        emitOptions();
        final var triple = tripleInQuadToProto(subject, predicate, object);
        rowBuffer.appendMessage().setStatementAdd(triple).getSerializedSize();
    }

    @Override
    public void deleteTriple(TNode subject, TNode predicate, TNode object) {
        emitOptions();
        final var triple = tripleInQuadToProto(subject, predicate, object);
        rowBuffer.appendMessage().setStatementDelete(triple).getSerializedSize();
    }

    @Override
    public void transactionStart() {
        emitOptions();
        // TODO: replace with .appendMessage() in when we introduce RowBuffer fully
        rowBuffer.add(ROW_TX_START);
    }

    @Override
    public void transactionCommit() {
        emitOptions();
        // TODO: replace with .appendMessage() in when we introduce RowBuffer fully
        rowBuffer.add(ROW_TX_COMMIT);
    }

    @Override
    public void transactionAbort() {
        emitOptions();
        // TODO: replace with .appendMessage() in when we introduce RowBuffer fully
        rowBuffer.add(ROW_TX_ABORT);
    }

    @Override
    public void addNamespace(String name, TNode iriValue, TNode graph) {
        final var namespace = encodeNamespace(name, iriValue, graph);
        rowBuffer.appendMessage().setNamespaceAdd(namespace).getSerializedSize();
    }

    @Override
    public void deleteNamespace(String name, TNode iriValue, TNode graph) {
        final var namespace = encodeNamespace(name, iriValue, graph);
        rowBuffer.appendMessage().setNamespaceDelete(namespace).getSerializedSize();
    }

    private RdfPatchNamespace encodeNamespace(String name, TNode iriValue, TNode graph) {
        emitOptions();
        final var namespace = RdfPatchNamespace.newInstance().setName(name);
        this.currentNsBase = namespace;
        if (iriValue != null) {
            this.currentTerm = SpoTerm.NAMESPACE;
            converter.nodeToProto(getNodeEncoder(), iriValue);
        }
        if (graph != null) {
            this.currentGraphBase = namespace;
            this.currentTerm = SpoTerm.GRAPH;
            this.graphNodeToProtoWrapped(graph);
        }
        return namespace;
    }

    @Override
    public void header(String key, TNode value) {
        emitOptions();
        final var header = RdfPatchHeader.newInstance().setKey(key);
        this.currentHeaderBase = header;
        this.currentTerm = SpoTerm.HEADER;
        converter.nodeToProto(getNodeEncoder(), value);
        rowBuffer.appendMessage().setHeader(header).getSerializedSize();
    }

    @Override
    public void punctuation() {
        emitOptions();
        if (options.getStreamType() != PatchStreamType.PUNCTUATED) {
            throw new RdfProtoSerializationError("Punctuation is not allowed in this stream type.");
        }
        // TODO: replace with .appendMessage() in when we introduce RowBuffer fully
        rowBuffer.add(ROW_PUNCTUATION);
    }

    private void emitOptions() {
        if (hasEmittedOptions) {
            return;
        }

        hasEmittedOptions = true;
        rowBuffer.appendMessage().setOptions(options).getSerializedSize();
    }
}
